runtime/fiber.c

```diff
@@ -102,10 +102,24 @@ struct stack_info** caml_alloc_stack_cache (void)
 
 Caml_inline struct stack_info* alloc_for_stack (mlsize_t wosize)
 {
-  size_t len = sizeof(struct stack_info) +
-               sizeof(value) * wosize +
-               8 /* for alignment to 16-bytes, needed for arm64 */ +
-               sizeof(struct stack_handler);
+  size_t stack_len = sizeof(struct stack_info) + sizeof(value) * wosize;
+  size_t len;
+
+  /* Some platforms require 16-byte alignment of the stack, so we need to
+     ask for a bit more memory to make sure that
+       caml_round_up(allocated stack base + stack_len, 16) + sizeof handler
+     will fit the allocated space.
+
+     When using mmap, we can rely upon the stack base being page-aligned
+     and thus aligned to a 16 byte boundary, and can round up here;
+     otherwise we need to always ask for 16 more bytes in order to cope with
+     all alignment possibilities. */
```

Comment:
The "subtle" thing here which makes the logic complicated is that what needs to be aligned to a 16-byte boundary is not the base address of the stack, but its top, immediately before the `stack_handler` data. (This answers the naive question)

So even if we would use `caml_stat_alloc_aligned_noexc`, there would still be a need to round up the allocation size. (Nitpick: prior to that, this function would need to be modified to use the nearest power of two larger or equal to the alignment required, rather than `Page_size`.)

Reply:
> The "subtle" thing here which makes the logic complicated is that what needs to be aligned to a 16-byte boundary is not the base address of the stack, but its top, immediately before the stack_handler data. 

I see, thanks! Can you make this clear in the comment? (Currently ou use say "16-byte alignment of the stack", which is not clear for people like me who know nothing about this stuff.)

Reply:
Comment updated.